File: app/frontend/src/app/services/backend.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { EventOffer } from '../models/EventOffer';

@Injectable({
  providedIn: 'root'
})
export class BackendService {

  baseUrl = 'http://localhost:8080/api/';

  constructor(private http: HttpClient) { }

  // ========== USER ==========
  register(user: any) {
    return this.http.post(`${this.baseUrl}users/register`, user);
  }

  login(username: string, password: string) {
    return this.http.post(`${this.baseUrl}users/login?username=${username}&password=${password}`, {});
  }

  updateUser(id: number, updated: any) {
    return this.http.post(`${this.baseUrl}users/${id}/update`, updated);
  }

  changePassword(id: number, newPassword: string) {
    return this.http.post(`${this.baseUrl}users/${id}/password?newPassword=${newPassword}`, {});
  }

  // ========== EVENTS ==========
  getEvents() {
    return this.http.get(`${this.baseUrl}events`);
  }

  getEvent(id: number) {
    return this.http.get(`${this.baseUrl}events/${id}`);
  }

  addEvent(event: any) {
    return this.http.post(`${this.baseUrl}events`, event);
  }

  // ========== OFFERS ==========
  getOffers() {
    return this.http.get(`${this.baseUrl}offers`);
  }

  getOffer(id: number) {
    return this.http.get(`${this.baseUrl}offers/${id}`);
  }

  // ========== PROMOTIONS ==========
  getPromotions() {
    return this.http.get(`${this.baseUrl}promotions`);
  }

  // ========== PHOTOS ==========
  getPhoto(eventId: number) {
    return this.http.get(`${this.baseUrl}photos/${eventId}`);
  }

  // ========== COMMENTS & RATINGS ==========
  addComment(eventId: number, comment: string, rating: number) {
    return this.http.post(`${this.baseUrl}events/${eventId}/comment?comment=${encodeURIComponent(comment)}&rating=${rating}`, {});
  }

  // ========== CART / SCHEDULING ==========
  addToCart(userId: number, offerId: number, numberOfPeople: number, date: Date) {
    const dateString = date instanceof Date ? date.toISOString().split('T')[0] : date;
  
    const params = new HttpParams()
      .set('offerId', offerId)
      .set('numberOfPeople', numberOfPeople)
      .set('date', dateString);
  
    return this.http.post(`${this.baseUrl}users/${userId}/cart`, {}, { params });
  }
  

  getCart(userId: number) {
    return this.http.get(`${this.baseUrl}users/${userId}/cart`);
  }

  removeFromCart(eventId: number) {
    const url = `${this.baseUrl}cart/${eventId}/delete`;
    return this.http.post(url, {}); 
  }
  
  confirmCart(userId: number) {
    return this.http.post(`${this.baseUrl}users/${userId}/cart/confirm`, {});
  }

  // ========== ORGANIZER ACTIONS ==========
  approveEvent(eventId: number) {
    return this.http.post(`${this.baseUrl}organizer/events/${eventId}/approve`, {});
  }

  declineEvent(eventId: number) {
    return this.http.post(`${this.baseUrl}organizer/events/${eventId}/decline`, {});
  }

  getEventsByStatus(statusId: number) {
    return this.http.get(`${this.baseUrl}events/status/${statusId}`);
  }

  addPhoto(base64File: string) {
    return this.http.post(`${this.baseUrl}photos`, { file: base64File });
  }
  
  addEventOffer(eventOffer: EventOffer) {
    return this.http.post(`${this.baseUrl}offers`, eventOffer);
  }
  

  // ========== NOTIFICATIONS ==========
  getNotifications(userId: number) {
    return this.http.get(`${this.baseUrl}users/${userId}/notifications`);
  }

  markNotificationAsRead(notificationId: number) {
    return this.http.post(`${this.baseUrl}notifications/${notificationId}/read`, {});
  }
}

File: PKI-android/app/src/main/java/com/example/pki/service/BackendService.java
package com.example.pki.service;

import com.example.pki.model.*;
import java.util.Date;
import java.util.List;

import retrofit2.Call;
import retrofit2.http.*;

public interface BackendService {

    // ========== USER ==========
    @POST("users/register")
    Call<Void> register(@Body User user);

    @POST("users/login")
    Call<User> login(@Query("username") String username, @Query("password") String password);

    @POST("users/{id}/update")
    Call<Void> updateUser(@Path("id") int id, @Body User updated);

    @POST("users/{id}/password")
    Call<Void> changePassword(@Path("id") int id, @Query("newPassword") String newPassword);

    // ========== EVENTS ==========
    @GET("events")
    Call<List<Event>> getEvents();

    @GET("events/{id}")
    Call<Event> getEvent(@Path("id") int id);

    @POST("events")
    Call<Void> addEvent(@Body Event event);

    // ========== OFFERS ==========
    @GET("offers")
    Call<List<EventOffer>> getOffers();

    @GET("offers/{id}")
    Call<EventOffer> getOffer(@Path("id") int id);

    @POST("offers")
    Call<Void> addEventOffer(@Body EventOffer eventOffer);

    // ========== PROMOTIONS ==========
    @GET("promotions")
    Call<List<Promotion>> getPromotions();

    // ========== PHOTOS ==========
    @GET("photos/{eventId}")
    Call<Photo> getPhoto(@Path("eventId") int eventId);

    @POST("photos")
    Call<Photo> addPhoto(@Body Photo photo);

    // ========== COMMENTS & RATINGS ==========
    @POST("events/{eventId}/comment")
    Call<Void> addComment(@Path("eventId") int eventId,
                          @Query("comment") String comment,
                          @Query("rating") int rating);

    // ========== CART / SCHEDULING ==========
    @POST("users/{userId}/cart")
    Call<Void> addToCart(@Path("userId") int userId,
                         @Query("offerId") int offerId,
                         @Query("numberOfPeople") int numberOfPeople,
                         @Query("date") String date); // date u ISO formatu yyyy-MM-dd

    @GET("users/{userId}/cart")
    Call<List<Event>> getCart(@Path("userId") int userId);

    @POST("cart/{eventId}/delete")
    Call<Void> removeFromCart(@Path("eventId") int eventId);

    @POST("users/{userId}/cart/confirm")
    Call<Void> confirmCart(@Path("userId") int userId);

    // ========== ORGANIZER ACTIONS ==========
    @POST("organizer/events/{eventId}/approve")
    Call<Void> approveEvent(@Path("eventId") int eventId);

    @POST("organizer/events/{eventId}/decline")
    Call<Void> declineEvent(@Path("eventId") int eventId);

    @GET("events/status/{statusId}")
    Call<List<Event>> getEventsByStatus(@Path("statusId") int statusId);

    // ========== NOTIFICATIONS ==========
    @GET("users/{userId}/notifications")
    Call<List<Notification>> getNotifications(@Path("userId") int userId);

    @POST("notifications/{notificationId}/read")
    Call<Void> markNotificationAsRead(@Path("notificationId") int notificationId);
}
